src/app/qgsmaptoolselectutils.cpp

```diff
@@ -335,3 +344,383 @@ QgsFeatureIds QgsMapToolSelectUtils::getMatchingFeatures( QgsMapCanvas *canvas,
+    r->stopRender( data->context );
+  return filterIds( newSelectedFeatures, data->existingSelection, data->selectBehavior );
+}
+
+void QgsMapToolSelectUtils::QgsMapToolSelectMenuActions::onSearchFinished()
+{
+  if ( mFutureWatcher && !mFutureWatcher->isFinished() )
+    return;
+
+  mAllFeatureIds = mFutureWatcher->result();
+  mActionChooseAll->setText( textForChooseAll( mAllFeatureIds.size() ) );
+  if ( !mAllFeatureIds.isEmpty() )
+    connect( mActionChooseAll, &QAction::hovered, this, &QgsMapToolSelectMenuActions::highlightFeatures );
+  else
+    mActionChooseAll->setEnabled( false );
+  if ( mAllFeatureIds.count() > 1 )
+    populateChooseOneMenu( mAllFeatureIds );
+}
+
+
+QString QgsMapToolSelectUtils::QgsMapToolSelectMenuActions::textForChooseAll( qint64 featureCount ) const
+{
+  if ( featureCount == 0 || featureCount == 1 )
+  {
+    switch ( mBehavior )
+    {
+      case QgsVectorLayer::SetSelection:
+        return tr( "Select" );
+        break;
+      case QgsVectorLayer::AddToSelection:
+        return tr( "Add to Selection" );
+        break;
+      case QgsVectorLayer::IntersectSelection:
+        return tr( "Intersect with Selection" );
+        break;
+      case QgsVectorLayer::RemoveFromSelection:
+        return tr( "Remove from Selection" );
+        break;
+    }
+  }
+
+  QString featureCountText;
+  if ( featureCount < 0 )
+    featureCountText = QStringLiteral( "Searching..." );
+  else
+    featureCountText = QString::number( featureCount );
+
+  switch ( mBehavior )
+  {
+    case QgsVectorLayer::SetSelection:
+      return tr( "Select All (%1)" ).arg( featureCountText );
+      break;
+    case QgsVectorLayer::AddToSelection:
+      return tr( "Add All to Selection (%1)" ).arg( featureCountText );
+      break;
+    case QgsVectorLayer::IntersectSelection:
+      return tr( "Intersect All with Selection (%1)" ).arg( featureCountText );
+      break;
+    case QgsVectorLayer::RemoveFromSelection:
+      return tr( "Remove All from Selection (%1)" ).arg( featureCountText );
+      break;
+  }
+
+  return QString();
+}
+
+QString QgsMapToolSelectUtils::QgsMapToolSelectMenuActions::textForChooseOneMenu() const
+{
+  switch ( mBehavior )
+  {
+    case QgsVectorLayer::SetSelection:
+      return tr( "Select One" );
+      break;
+    case QgsVectorLayer::AddToSelection:
+      return tr( "Add One to Selection" );
+      break;
+    case QgsVectorLayer::IntersectSelection:
+      return tr( "Intersect One with Selection" );
+      break;
+    case QgsVectorLayer::RemoveFromSelection:
+      return tr( "Remove One from Selection" );
+      break;
+  }
+
+  return QString();
+}
+
+
+void QgsMapToolSelectUtils::QgsMapToolSelectMenuActions::populateChooseOneMenu( const QgsFeatureIds &ids )
+{
+  QgsFeatureIds displayedFeatureIds;
+
+  QgsFeatureIds::ConstIterator it = ids.constBegin();
+  while ( displayedFeatureIds.count() <= 20 && it != ids.constEnd() ) //for now hardcoded, but maybe define a settings for this
```

Comment:
This same threshold should be set for setLimit on the feature request getting the features -- otherwise potentially we are fetching a bunch of features which we'll never use

Reply:
I am not sure, if the user wants "all", we have to request all. The limit is only if the user wants one.

Reply:
Ah, I missed that!